---
title: "R Notebook"
output: html_notebook
---

```{r}
projectPath <- file.path("C:/Users/ctorroja/Desktop/scMaster/2024_BioInfoUAMMaster/SingleCell/Stuart2019")
outputPath <- file.path(projectPath,"output")
dir.create(outputPath)

dataPath <- file.path(projectPath,"GSE128639_10XGenomics/outs")

setwd(projectPath)

```

```{r eval=FALSE, include=FALSE}
BiocManager::install(c("scran","scuttle","biomaRt"), type = "binary")
install.packages(c("Seurat","tidyverse","gtools"), type = "binary")
```

```{r}
library(Seurat)
library(scuttle)
library(scran)
library(biomaRt)
library(tidyverse)
library(gtools)
library(Matrix)
library(RColorBrewer)
library(ggridges)
library(UpSetR)

color.list <- RColorBrewer::brewer.pal(12,"Paired")
color.list <- c(color.list,RColorBrewer::brewer.pal(12,"Set3"))
```

# Load CellRanger Quantification

```{r eval=FALSE, include=FALSE}
counts <- Read10X(
  file.path(dataPath,"filtered_feature_bc_matrix"),
  gene.column = 1)

selectedCells <- sample(x = colnames(counts$`Gene Expression`),
                        size = 5000,
                        replace = F)

counts <- lapply(counts, function (x) {return(x[,selectedCells])})

counts <- do.call(rbind,counts[c(1,3,2)])

features <- read_tsv(file.path(dataPath,"filtered_feature_bc_matrix","features.tsv.gz"))

dir.create(file.path(dataPath,"sampled_feature_bc_matrix"))

writeMM(obj = countsnew, 
        file = gzfile(file.path(dataPath,"sampled_feature_bc_matrix","matrix.mtx.gz")))

file.copy(
  from = file.path(dataPath,"filtered_feature_bc_matrix","features.tsv.gz"),
  to = file.path(dataPath,"sampled_feature_bc_matrix","features.tsv.gz"))

write_tsv(data.frame(selectedCells),
          file = gzfile(file.path(dataPath,"sampled_feature_bc_matrix","barcodes.tsv.gz")),
          col_names = F)

counts <- countsnew <- NULL
gc()
```

```{r}
counts <- Read10X(file.path(dataPath,"sampled_feature_bc_matrix"),
                  gene.column = 1)
lapply(counts, dim)
```

# Create SingleCellExperiment

```{r}
sce <- SingleCellExperiment(
  assays = list(counts = counts$`Gene Expression`),
  mainExpName = "RNA"
)

adt <- SingleCellExperiment(
  assays = list(counts = counts$`Antibody Capture`)
)

hto <- SingleCellExperiment(
  assays = list(counts = counts$Custom)
)

altExps(sce) <- list(ADT = adt, HTO = hto)

```

# Load Gene Metadata from BioMart

```{r}

mart <- useEnsembl(biomart = 'genes',
                   dataset = "hsapiens_gene_ensembl",
                   version = 99)

genesMetadata <- getBM(
  attributes = c(
    "external_gene_name"
    ,"hgnc_symbol"
    ,"ensembl_gene_id"
    ,"gene_biotype"
    ,"chromosome_name"
    ,"start_position"
    ,"end_position","strand"
    ,"description"
  ),
  uniqueRows=T, 
  mart = mart,
  verbose = F)

sum(duplicated(genesMetadata$ensembl_gene_id))

genesMetadata <- genesMetadata %>%
  filter(!grepl("^CHR_",chromosome_name))

sum(duplicated(genesMetadata$ensembl_gene_id))

genesMetadata[duplicated(genesMetadata$ensembl_gene_id),]

genesMetadata <- genesMetadata %>%
  filter(!duplicated(genesMetadata$ensembl_gene_id))

sum(duplicated(genesMetadata$external_gene_name))

genesMetadata <- genesMetadata %>%
  mutate(uniq_name = make.names(external_gene_name,unique = T))

sum(duplicated(genesMetadata$uniq_name))

dim(genesMetadata)
```

```{r}

genesMetadataInExperiment <- genesMetadata %>%
  filter(ensembl_gene_id %in% rownames(sce))

genesMetadataInExperiment <- genesMetadataInExperiment %>%
  mutate(geneId = ensembl_gene_id) %>% 
  column_to_rownames("geneId")

rowData(sce)$ensembl_gene_id <- rownames(sce)

df <- as.data.frame(rowData(sce))

df <- df %>%
  left_join(genesMetadataInExperiment, by = c("ensembl_gene_id")) %>%
  mutate(uniq_name = ifelse(
    is.na(uniq_name),ensembl_gene_id,uniq_name)) %>%
  as.data.frame()

rownames(df) <- df$ensembl_gene_id

rowData(sce) <- DataFrame(df)

rowData(sce)$strand <- NULL
```

# Annotate Genes

```{r}
# df <- as.data.frame(rowData(sce))

rownames(sce) <- rowData(sce)$uniq_name

# genesMetadataInExperiment[match(rownames(sce),genesMetadataInExperiment$ensembl_gene_id),"uniq_name"]

HBBGenes <- genesMetadataInExperiment %>%
  filter(grepl("hemoglobin",description,ignore.case = T))

IGGenes <- genesMetadataInExperiment %>%
  filter(grepl("^IG_\\w*",gene_biotype,perl=T,ignore.case = T))

TRGenes <- genesMetadataInExperiment %>%
  filter(grepl("^TR_\\w*",gene_biotype,perl=T,ignore.case = T))

MTGenes <- genesMetadataInExperiment %>%
  filter(chromosome_name == "MT")

RBRGenes <- genesMetadataInExperiment %>% 
  filter(gene_biotype == "rRNA")

G2MGenes <- genesMetadataInExperiment %>%
  filter(external_gene_name %in% cc.genes$g2m.genes)

SGenes <- genesMetadataInExperiment %>%
  filter(external_gene_name %in% cc.genes$s.genes)

```

# Compute QC Stats

```{r}
sce <- sce %>%
  addPerCellQC(
    subsets = list(
      MT = MTGenes$uniq_name
      # ,RBR = RBRGenes$uniq_name
      ,HBB = HBBGenes$uniq_name
      ,IG = IGGenes$uniq_name
      ,TR = TRGenes$uniq_name
    )
    ,percent.top = c(50)
    ,use_altexps = T
  )

sce <- addPerFeatureQC(
  sce
)

# rowData(sce)$is_feature_control <- rownames(rowData(sce)) %in% c(MTGenes$uniq_name,RBRGenes$uniq_name,HBBGenes$uniq_name,IGGenes$uniq_name,TRGenes$uniq_name)

altExp(sce,"ADT") <- addPerFeatureQC(
  altExp(sce,"ADT")
)
altExp(sce,"ADT") <- addPerCellQC(
  altExp(sce,"ADT")
)

altExp(sce,"HTO") <- addPerFeatureQC(
  altExp(sce,"HTO")
)
altExp(sce,"HTO") <- addPerCellQC(
  altExp(sce,"HTO")
)
```

# Filtering

```{r}
df <- as.data.frame(colData(sce))
```

## Filtering by sequencing depth

```{r}
minDepth <- 800
maxDepth <- 50000
```

To minimize the amount of low quality cells and improve posterior normalization and analysis we have applied a minimum of normalized counts per cell of `r minDepth` and a maximum of `r maxDepth`.

```{r }

### Adjust minDepth
df %>%
  ggplot(aes(x=sum)) +
    geom_density(stat = "density",alpha=.5) +
    scale_fill_manual(values = color.list) +
    geom_vline(xintercept = minDepth, color="red") +
    geom_vline(xintercept = maxDepth, color="red") +
    coord_cartesian(xlim=c(0,maxDepth*1.1)) +
    ggtitle("Sequencing Depth Density Plot") +
    theme(legend.position = "none")

df %>%
  arrange(sum) %>%
  summarise(CumSum_Counts=cumsum(sum),
            total_counts=sum) %>%
  ggplot(aes(x=total_counts,y=CumSum_Counts)) +
  geom_point() +
  geom_vline(xintercept = minDepth, color="red") +
  geom_vline(xintercept = maxDepth, color="red") +
  coord_cartesian(xlim=c(0,maxDepth*1.1)) +
  ggtitle("Sequencing Depth Density Plot") +
  theme(legend.position = "none")

```


```{r}
########### Filter By Total Counts ############
sce$filter_by_total_counts <- sce$sum > minDepth & sce$sum < maxDepth

table(sce@colData[,c("filter_by_total_counts")])

```

## Filtering by genes detected

```{r}
minGenesDetected <- 400
```


we have applied also a minimum gene detection filter of `r minGenesDetected` genes.

```{r }

df %>%
  ggplot(aes(x=detected)) +
  geom_density(stat="density",alpha=.5) +
  scale_fill_manual(values = color.list) +
  geom_vline(xintercept = minGenesDetected, color="red") +
  ggtitle("Genes Detected Density Plot") +
  theme(legend.position = "none")

df %>%
  arrange(detected) %>%
  summarise(CumSum_Features=cumsum(detected),
            total_features=detected) %>%
  ggplot(aes(x=total_features,y=CumSum_Features)) +
  geom_point() +
  geom_vline(xintercept = minGenesDetected, color="red") +
  ggtitle("Genes Detected Cummulative Plot") +
  theme(legend.position = "none")

```


```{r}
########## Filter By Detected Genes ###########
sce$filter_by_expr_features <- sce@colData$detected > minGenesDetected

table(sce$filter_by_expr_features)

```

## Filter by MT content

```{r}

maxMT <- 15

```

A maximum of `r maxMT`% of MT content has been used to reduce the amount of low quality cells.

```{r }
df %>%
  ggplot(aes(x=subsets_MT_percent,y="Sample")) +
  geom_density_ridges(aes(fill="Sample"),alpha=0.5) +
  scale_fill_manual(values = color.list) +
  geom_jitter(height = 0.2, alpha = 0.01) +
  geom_vline(xintercept = maxMT, color="red") +
  ggtitle("MT Content Density Plot") +
  theme(legend.position = "none")

```

```{r}
########## Filter By Detected Genes ###########
sce$filter_by_MT_features <- sce$subsets_MT_percent < maxMT

table(sce$filter_by_MT_features)

```


## Filter by Gene Expression Complexity

```{r}

maxPct <- 65

```

Cells with high levels of reads in just a few genes have been filtered. `r maxPct`% of reads in the top 50 genes.

```{r }

df %>%
  ggplot(aes(x=percent.top_50,y="Sample")) +
  geom_density_ridges(aes(fill="Sample")) +
  geom_jitter(height = 0.2, alpha = 0.01) +
  geom_vline(xintercept = maxPct, color="red") +
  scale_fill_manual(values = color.list) +
  ggtitle("UMIs in Top 50 Genes Density Plot") +
  theme(legend.position = "none")

```

```{r}
########## Filter By Detected Genes ###########
sce$filter_by_Top50 <- sce$percent.top_50 < maxPct

table(sce$filter_by_Top50)

```

## Filter by HBB Gene Set Expression

```{r}

maxHBB <- 0.2

```

In case we've got some erythrocytes or excessive contamination from their lysis we have removed cells with more than `r maxHBB`% reads in HBB genes.

```{r }

df %>%
  ggplot(aes(x=subsets_HBB_percent,y="Sample")) +
  geom_density_ridges(aes(fill="Sample")) +
  geom_jitter(height = 0.2, alpha = 0.01) +
  geom_vline(xintercept = maxHBB, color="red") +
  scale_fill_manual(values = color.list) +
  ggtitle("HBB Content Density Plot") +
  theme(legend.position = "none")
```

```{r}
########## Filter By Detected Genes ###########
sce$filter_by_HBB_features <- sce$subsets_HBB_percent < maxHBB

table(sce$filter_by_HBB_features)
```

## HTO Content

```{r}

minHTO <- 50

```

We have filtered out also cells with less than `r minHTO` UMIs in the hashtags to improve sample calling.

```{r }
df %>%
  ggplot(aes(x=altexps_HTO_sum)) +
    geom_density(stat = "density",aes(fill = "Sample"),alpha=.5) +
    scale_fill_manual(values = color.list) +
    geom_vline(xintercept = minHTO, color="red") +
    coord_cartesian(xlim=c(0,maxDepth*1.1)) +
    ggtitle("HashTag Counts Density Plot") +
    theme(legend.position = "none")

df %>%
  arrange(altexps_HTO_sum) %>%
  summarise(CumSum_Counts=cumsum(altexps_HTO_sum),
            Sample="Sample",HTO_counts=altexps_HTO_sum) %>%
  ggplot(aes(x=HTO_counts,y=CumSum_Counts,color=Sample)) +
  geom_point() +
  geom_vline(xintercept = minHTO, color="red") +
  ggtitle("HashTag Counts Cummulative Plot") +
  theme(legend.position = "none")

```

```{r}
########## Filter By Detected Genes ###########
sce$filter_by_HTO_features <- sce$altexps_HTO_sum > minHTO

table(sce@colData[,c("filter_by_HTO_features")])

```

## ADT Content

```{r}

minADT <- 100

```

We have filtered out also cells with less than `r minHTO` UMIs in the hashtags to improve sample calling.

```{r }
df %>%
  ggplot(aes(x=altexps_ADT_sum)) +
  geom_density(stat = "density",aes(fill = "Sample"),alpha=.5) +
  scale_fill_manual(values = color.list) +
  geom_vline(xintercept = minADT, color="red") +
  coord_cartesian(xlim=c(0,maxDepth*1.1)) +
  ggtitle("Antibody Counts Density Plot") +
  theme(legend.position = "none")

df %>%
  arrange(altexps_ADT_sum) %>%
  summarise(CumSum_Counts=cumsum(altexps_ADT_sum),
            Sample="Sample",ADT_counts=altexps_ADT_sum) %>%
  ggplot(aes(x=ADT_counts,y=CumSum_Counts,color=Sample)) +
  geom_point() +
  geom_vline(xintercept = minADT, color="red") +
  ggtitle("Antibody Counts Cummulative Plot") +
  theme(legend.position = "none")

```

```{r}
########## Filter By Detected Genes ###########
sce$filter_by_ADT_features <- sce$altexps_ADT_sum > minADT

table(sce$filter_by_ADT_features)
```

## Filtering Summary

```{r }

######### Set Manual Filtering ##########
sce$manualFilter <- (
  sce$filter_by_expr_features &
  sce$filter_by_total_counts &
  sce$filter_by_MT_features &
  sce$filter_by_HBB_features &
  sce$filter_by_Top50 &
  sce$filter_by_HTO_features &
  sce$filter_by_ADT_features
)

########### %DropOut vs MT Features ###########
df <- as.data.frame(colData(sce)) %>%
  dplyr::select(sum,detected,
                filter_by_expr_features,
                filter_by_total_counts,
                filter_by_Top50,
                filter_by_HTO_features,
                filter_by_ADT_features,
                filter_by_MT_features,
                filter_by_HBB_features,
                manualFilter)

df <- df %>%
  pivot_longer(cols = contains("filter",ignore.case = T),
               names_to = "Filter",
               values_to = "PASS")

df %>%
  ggplot(aes(x=log10(sum),y=detected)) +
  geom_point(aes(color=PASS)) +
  facet_wrap("Filter")

```

```{r}
table(sce$manualFilter)
```

```{r}
df <- as.data.frame(colData(sce)) %>%
  dplyr::select(contains("filter_by")) %>%
  mutate_all(.funs = ~ !.) %>%
  mutate_all(.funs = as.integer)

upset(df)
```

# Filter Cells

```{r}
scef <- sce[,sce$manualFilter]
scef
```


# Normalization

## Normalization Factors

```{r}
clusters <- scef %>%
  quickCluster(min.size = 200)

# sce <- sce %>% computePooledFactors()
pooledSizeFactors <- scef %>%
  pooledSizeFactors(
    clusters = clusters,
    assay.type = "counts"
  )

# sce <- sce %>% computeLibraryFactors()
librarySizeFactors <- scef %>%
  librarySizeFactors()
```

```{r}
plot(pooledSizeFactors,librarySizeFactors)
```

# Normalize Data

```{r}
scef <- logNormCounts(scef,size.factors = pooledSizeFactors)

# Centered Log Ratio
clr <- function(x) {
          return(log1p(x = x / (expm1(x = sum(log1p(x = x[x > 0]), na.rm = TRUE) / length(x = x)))))
}

# A special
assay(altExp(scef,"HTO"),"logcounts") <- clr(assay(altExp(scef,"HTO"),"counts"))

assay(altExp(scef,"ADT"),"logcounts") <- clr(assay(altExp(scef,"ADT"),"counts"))
# 
# ModelNegativeADTnorm(cell_protein_matrix = assay(altExp(sce,"adt"),slot = "counts"), 
#                                 denoise.counts = TRUE,
#                                 use.isotype.control = FALSE)
```

## Explore QCs

```{r, fig.asp=1.8}

scnames <- sample(colnames(scef),200)

dc <- as.data.frame(as.matrix(assay(scef,"logcounts"))[,scnames]) %>%
  mutate(Gene = rownames(as.matrix(assay(scef,"logcounts")))) %>%
  pivot_longer(cols = -Gene,names_to = "Cell",values_to = "Exprs")

dc <- dc %>% filter(Exprs>0)
# dc$Sample <- factor(colData(sce)[dc$Cell,"Sample"])
dc$Sample <- "BMMCells"
dc$Features <- colData(scef)[dc$Cell,"detected"]

# Relevel Cell by Sample Type
dc$Cell <- factor(dc$Cell,levels=unique(dc$Cell[order(dc$Sample,dc$Cell)]))

g1 <- ggplot(dc,aes(y=Cell,x=Exprs)) +
  geom_density_ridges(aes(fill=Sample)) +
  xlab("Gene Expression") +
  scale_fill_manual(values = color.list) +
  theme(legend.position = "top",legend.direction = "horizontal"
        ,axis.text = element_text(size = 6))

g2 <- ggplot(dc,aes(x=Cell,fill=Sample)) +
  geom_bar(stat = "count") +
  ylab("Detected Genes") +
  scale_fill_manual(values = color.list) +
  coord_flip() +
  theme(legend.position = "none",legend.direction = "horizontal"
        ,axis.text.y = element_blank(),axis.title.y = element_blank())

g3 <- ggplot(dc,aes(x=Cell,y=Exprs,fill=Sample)) +
  geom_bar(stat = "sum") +
  ylab("Total Counts") +
  scale_fill_manual(values = color.list) +
  coord_flip() +
  theme(legend.position = "none",legend.direction = "horizontal"
        ,axis.text.y = element_blank(),axis.title.y = element_blank())

g4 <- ggplot(dc %>% filter(Exprs > 0) %>% group_by(Cell,Sample) %>% summarise(Mean = mean(Exprs)),aes(x=Cell,y=Mean,fill=Sample)) +
  geom_bar(stat = "identity") +
  ylab("Mean Gene Exp") +
  scale_fill_manual(values = color.list) +
  coord_flip() +
  theme(legend.position = "none",legend.direction = "horizontal"
        ,axis.text.y = element_blank(),axis.title.y = element_blank())

p <- g2 + g3 + g4

g1 + p

```

# Create Seurat Object

```{r}
myGenes <- !rownames(scef) %in% c(MTGenes$uniq_name,
                                 HBBGenes$uniq_name,
                                 RBRGenes$uniq_name)
myCells <- scef$manualFilter

scef <- scef[myGenes,]

seuratSC <- as.Seurat(scef
                      , counts = "counts"
                      , data = "logcounts"
                      , project = "Stuart2019"
)

# seuratSC <- CreateSeuratObject(
#   counts = assay(scef,"counts"),
#   data = assay(scef,"logcounts"))
# 
# seuratSC <- seuratSC %>%
#   AddMetaData(
#     metadata = as.data.frame(colData(scef))
#     )
# 
# seuratSC@assays$ADT <- CreateAssayObject(
#   key = "adt_",
#   counts = assay(altExp(scef,"ADT"),"counts")
# )
# 
# seuratSC@assays$HTO <- CreateAssay5Object(
#   key = "hto_",
#   counts = assay(altExp(scef,"HTO"),"counts"),
#   data = assay(altExp(scef,"HTO"),"logcounts")
# )

seuratSC
```

```{r}
Seurat::Assays(seuratSC)
```

```{r}
DefaultAssay(seuratSC)
```

## Normalize HTO Data and Demultiplex

```{r}
DefaultAssay(seuratSC) <- "HTO"

seuratSC <- seuratSC %>%
  NormalizeData(
    assay = "HTO",
    normalization.method = "CLR",
    margin = 1)

seuratSC <- seuratSC %>%
  HTODemux(
    assay = "HTO",
    positive.quantile = 0.95)

table(seuratSC$HTO_classification.global)
```

```{r}
Idents(seuratSC) <- "hash.ID"
```

```{r fig.height=10}

seuratSC %>%
  RidgePlot(
    assay = "HTO",
    features = rownames(seuratSC[["HTO"]])[1:4],
    ncol = 2)

seuratSC %>%
  RidgePlot(
    assay = "HTO",
    features = rownames(seuratSC[["HTO"]])[5:10],
    ncol = 3)

```


```{r}
seuratSC %>%
  HTOHeatmap(
    assay = "HTO",
    ncells = 6000)
```


```{r message=FALSE, warning=FALSE}

myBCs <-  rownames(seuratSC@assays$HTO)[1:4]
pList <- list()
for (i1 in seq(length(myBCs)-1)) {
  for (i2 in seq(from=i1+1,to=length(myBCs))) {
    bc1 <- myBCs[i1]
    bc2 <- myBCs[i2]
    pList[[paste(bc1,bc2,sep="_")]] <- (FeatureScatter(subset(seuratSC,HTO_classification %in% c(bc1,paste(bc1,bc2,sep="_"),bc2)),feature1=bc1,feature2=bc2,group.by = "HTO_classification"))
  }
}
p <- lapply(pList,print)
```

## Normalize and Scale CITE-Seq/Ab-Seq Data

```{r}
DefaultAssay(seuratSC) <- "ADT"
```

```{r}
seuratSC <- seuratSC %>%
  NormalizeData(
    assay = "ADT",
    normalization.method = "CLR",
    margin = 2) %>%
  ScaleData()
```

```{r}
rownames(seuratSC)
```


```{r}
seuratSC %>%
  FeatureScatter(
    "CD3",
    "CD14",
    group.by = "HTO_classification.global"
  )
```

## Detect Doublets and remove them

```{r}
library(scDblFinder)
DefaultAssay(seuratSC) <- "RNA"

sce <- scDblFinder(as.SingleCellExperiment(seuratSC),knownDoublets = seuratSC$HTO_classification.global == "Doublet",)

seuratSC <- as.Seurat(sce
                      , counts = "counts"
                      , data = "logcounts")

```

```{r}
table(seuratSC$scDblFinder.class,seuratSC$HTO_classification.global)
```

```{r}
DefaultAssay(seuratSC) <- "ADT"
p1 <- seuratSC %>%
  FeatureScatter(
    "CD3",
    "CD14",
    group.by = "scDblFinder.class"
  )

p2 <- seuratSC %>%
  FeatureScatter(
    "CD3",
    "CD14",
    group.by = "HTO_classification.global"
  )

p1 + p2 + patchwork::plot_layout(ncol = 2)
```

## Remove Doublets

```{r}
seuratSC <- seuratSC[,seuratSC$scDblFinder.class == "singlet"]
```

## Normalize and Scale CITE-Seq/Ab-Seq Data

```{r}
DefaultAssay(seuratSC) <- "ADT"
```

```{r}
seuratSC <- seuratSC %>%
  NormalizeData(
    assay = "ADT",
    normalization.method = "CLR",
    margin = 2) %>%
  ScaleData()
```

## Normalize Transcriptome Data

```{r}
DefaultAssay(seuratSC) <- "RNA"
```

```{r}

seuratSC <- seuratSC %>%
  NormalizeData(
    assay = "RNA",
    normalization.method = "LogNormalize"
  ) %>%
  FindVariableFeatures(
    assay = "RNA",
    nfeatures = 2000,
    selection.method = "vst", # or mvp
    verbose = F) %>%
  ScaleData()

```

```{r}
seuratSC <- seuratSC %>%
  SCTransform(
    assay = "RNA",
    conserve.memory = T,
    variable.features.n = 2000,
    new.assay.name = "SCT"
  )
```

```{r}
# Get the list of most var genes from lognorm data
DefaultAssay(seuratSC) <- "RNA"
vgvst <- seuratSC %>% VariableFeatures(
  assay = "RNA",
  selection.method = "vst"
)

# Get the list of most var genes from SCTnorm data
DefaultAssay(seuratSC) <- "SCT"
vgsct <- seuratSC %>% VariableFeatures(
  assay = "SCT"
)

# Get the list of proteins from ADT data
vgadt <- rownames(seuratSC@assays$ADT)

sum(vgvst %in% vgsct)
```

```{r}
features <- c("nFeature_RNA","nCount_RNA","subsets_MT_percent","altexps_ADT_percent","altexps_HTO_percent","percent.top_50")

VlnPlot(subset(seuratSC,HTO_classification.global == "Singlet")
        ,features = features
        ,group.by = "HTO_classification"
        ,cols = color.list
        ,ncol = 4
        ,pt.size = 0
        ,combine = F
        )
```

```{r}
seuratSC <- seuratSC %>%
  RunPCA(
    assay = "RNA",
    npcs = 50,
    features = vgvst,
    reduction.name = "pca",
    reduction.key = "PC_",
    verbose = F
  ) %>%
  RunPCA(
    assay = "SCT",
    npcs = 50,
    features = vgsct,
    reduction.name = "SCTpca",
    reduction.key = "SCTPC_",
    verbose = F
  ) %>%
  RunPCA(
    assay = "ADT",
    npcs = 50,
    features = vgadt,
    reduction.name = "ADTpca",
    reduction.key = "ADTPC_",
    verbose = F
  )
```

## QC PCAs

### CC Module on PCAs

#### Compute Cell Cycle Scores

```{r }

this.s.genes <- rownames(seuratSC)[rownames(seuratSC) %in% SGenes$uniq_name]

this.g2m.genes <- rownames(seuratSC)[rownames(seuratSC) %in% G2MGenes$uniq_name]

```

```{r}
seuratSC <- seuratSC %>%
  CellCycleScoring(
    assay = "RNA",
    s.features = this.s.genes,
    g2m.features = this.g2m.genes) %>%
 AddModuleScore(
   assay = "RNA",
   features = list(cc.Score = c(this.g2m.genes,this.s.genes)),
   name = "CC.Score")
```

#### Compute Enrichment of CC genes in PC componets

```{r}
# Get All PC Embeddings
df <- Embeddings(seuratSC,reduction = "pca")

# Add Metadata
df <- cbind(df,seuratSC@meta.data)

# Sample Data for plotting purposes.
df <- df[sample(nrow(df),size = min(10000,length(Cells(seuratSC)))),]
```

```{r}
# Get list of all CC genes
ccGenesVars <- c(SGenes$uniq_name,G2MGenes$uniq_name)

# Filter to those in our single cell object
ccGenesVars <- ccGenesVars[ccGenesVars %in% VariableFeatures(seuratSC, assay = "RNA")]

# Compute enrichment of PC loadings of CC gene set
unloadings <- data.frame(
  cc = colMeans(abs(seuratSC@reductions$pca@feature.loadings[ccGenesVars,1:30]))/colMeans(abs(seuratSC@reductions$pca@feature.loadings[,1:30]))
  ) %>%
  mutate(PC = rownames(.)) %>%
  mutate(PC = factor(PC,levels = mixedsort(PC)))
```

```{r}
pairs(df[,paste0("PC_",1:30)]
      , row1attop = F
      , gap = 0.25
      , pch = 19
      , cex = 0.5
      , col = colorRamp2(
        breaks = c(min(df$CC.Score1),mean(df$CC.Score1),max(df$CC.Score1)),
        colors = c("blue","yellow","red"))(df$CC.Score1)
      )
 # RColorBrewer::brewer.pal(3,"Spectral")
```

```{r}
unloadings %>%
  ggplot(aes(x=PC,y=cc)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(angle = 90))

```

## Variability across PCs

```{r}
seuratSC %>%
  ElbowPlot(ndims = 30,reduction = "pca")
```

```{r}
seuratSC %>%
  ElbowPlot(ndims = 30,reduction = "ADTpca")
```

```{r}
seuratSC %>%
  ElbowPlot(ndims = 30,reduction = "SCTpca")
```

# Clustering and Dimensionality Reduction

```{r}
seuratSC <- seuratSC %>%
  FindNeighbors(
    assay = "RNA",
    reduction = "pca",
    dims = 1:20,
    features = vgvst,
    k.param = 20,
    force.recalc = T,
    graph.name = c("RNA_nn","RNA_snn"),
  ) %>%
  FindNeighbors(
    assay = "ADT",
    reduction = "ADTpca",
    dims = 1:20,
    features = vgadt,
    k.param = 20,
    force.recalc = T,
    graph.name = c("ADT_nn","ADT_snn"),
  ) %>%
  FindNeighbors(
    assay = "SCT",
    reduction = "SCTpca",
    dims = 1:20,
    features = vgsct,
    k.param = 20,
    force.recalc = T,
    graph.name = c("SCT_nn","SCT_snn"),
  )
```

```{r}
seuratSC <- seuratSC %>%
  FindClusters(
    assay = "RNA",
    graph.name = "RNA_snn",
    resolution = c(0.1,0.25,0.5,0.75,1),
  ) %>%
  FindClusters(
    assay = "ADT",
    graph.name = "ADT_snn",
    resolution = c(0.1,0.25,0.5,0.75,1),
  ) %>%
  FindClusters(
    assay = "SCT",
    graph.name = "SCT_snn",
    resolution = c(0.1,0.25,0.5,0.75,1),
  )
```


```{r}
seuratSC <- seuratSC %>%
  RunUMAP(
    assay = "RNA",
    reduction = "pca",
    dims = 1:16,
    min.dist = 0.3,
    spread = 1,
    n.neighbors = 30L,
    metric = "cosine", # euclidean
    reduction.name = "RNAumap",
    reduction.key = "RNAUMAP_"
  ) %>%
  RunUMAP(
    assay = "ADT",
    reduction = "ADTpca",
    dims = 1:16,
    min.dist = 0.3,
    spread = 1,
    n.neighbors = 30L,
    metric = "cosine", # euclidean
    reduction.name = "ADTumap",
    reduction.key = "ADTUMAP_"
  ) %>%
  RunUMAP(
    assay = "SCT",
    reduction = "SCTpca",
    dims = 1:16,
    min.dist = 0.3,
    spread = 1,
    n.neighbors = 30L,
    metric = "cosine", # euclidean
    reduction.name = "SCTumap",
    reduction.key = "SCTUMAP_"
  )
```

## Clustering based on LogNorm

```{r}
p1 <- DimPlot(seuratSC,
        reduction = "RNAumap",
        group.by = "RNA_snn_res.0.5",
        label = T,
        cols = color.list) +
  theme(legend.position = "none")

p2 <- DimPlot(seuratSC,
        reduction = "SCTumap",
        group.by = "RNA_snn_res.0.5",
        label = T,
        cols = color.list)+
  theme(legend.position = "none")

p3 <- DimPlot(seuratSC,
        reduction = "ADTumap",
        group.by = "RNA_snn_res.0.5",
        label = T,
        cols = color.list)

p1 + p2 + p3 + patchwork::plot_layout(ncol = 3)
```

## Clustering based on SCT

```{r}
p1 <- DimPlot(seuratSC,
        reduction = "RNAumap",
        group.by = "SCT_snn_res.0.5",
        label = T,
        cols = color.list) +
  theme(legend.position = "none")

p2 <- DimPlot(seuratSC,
        reduction = "SCTumap",
        group.by = "SCT_snn_res.0.5",
        label = T,
        cols = color.list)+
  theme(legend.position = "none")

p3 <- DimPlot(seuratSC,
        reduction = "ADTumap",
        group.by = "SCT_snn_res.0.5",
        label = T,
        cols = color.list)

p1 + p2 + p3 + patchwork::plot_layout(ncol = 3)
```

## Clustering based on ADT

```{r}
p1 <- DimPlot(seuratSC,
        reduction = "RNAumap",
        group.by = "ADT_snn_res.0.5",
        label = T,
        cols = color.list) +
  theme(legend.position = "none")

p2 <- DimPlot(seuratSC,
        reduction = "SCTumap",
        group.by = "ADT_snn_res.0.5",
        label = T,
        cols = color.list)+
  theme(legend.position = "none")

p3 <- DimPlot(seuratSC,
        reduction = "ADTumap",
        group.by = "ADT_snn_res.0.5",
        label = T,
        cols = color.list)

p1 + p2 + p3 + patchwork::plot_layout(ncol = 3)
```

## Different Resolutions

```{r}
df <- Embeddings(seuratSC, reduction = "RNAumap")
dnames <- colnames(df)
df <- cbind(df,seuratSC@meta.data)
```

```{r}
df %>%
  select(contains(c("RNAUMAP","RNA_snn"))) %>%
  pivot_longer(cols = contains("RNA_snn"),names_to = "Resolution",values_to = "Cluster") %>%
  ggplot(aes_string(x=dnames[1],y=dnames[2],color = "Cluster")) +
           geom_point() +
           scale_color_manual(values = color.list) +
           facet_wrap("Resolution") +
           theme_classic()
```

# Cluster annotation

```{r}
library(SingleR)
dbImm <- celldex::DatabaseImmuneCellExpressionData()
moImm <- celldex::MonacoImmuneData()

cellannot <- SingleR(
  clusters=NULL
  , test = Seurat::Assays(seuratSC,slot = "RNA")@data
  , ref = dbImm
  , labels = dbImm$label.fine
  , genes = "de"
  , quantile = 0.8
  , fine.tune = T
  , tune.thresh = 0.05
  , sd.thresh = 1
)
```

```{r}
colnames(cellannot) <- paste0("dbImm_",colnames(cellannot))
df <- cbind(df,cellannot)
```

```{r}
df %>%
  select(contains(c("RNAUMAP","dbImm"))) %>%
  pivot_longer(cols = contains("labels"),names_to = "Label",values_to = "CellType") %>%
  ggplot(aes_string(x=dnames[1],y=dnames[2],color = "CellType")) +
           geom_point() +
           scale_color_manual(values = color.list) +
           facet_wrap("Label") +
           theme_classic()
```

```{r}
cellannot <- SingleR(
  clusters=NULL
  , test = Seurat::Assays(seuratSC,slot = "RNA")@data
  , ref = moImm
  , labels = moImm$label.fine
  , genes = "de"
  , quantile = 0.8
  , fine.tune = T
  , tune.thresh = 0.05
  , sd.thresh = 1
)
```

```{r}
colnames(cellannot) <- paste0("moImm_",colnames(cellannot))
df <- cbind(df,cellannot)
```

```{r}
df %>%
  select(contains(c("RNAUMAP","moImm"))) %>%
  pivot_longer(cols = contains("labels"),names_to = "Label",values_to = "CellType") %>%
  ggplot(aes_string(x=dnames[1],y=dnames[2],color = "CellType")) +
           geom_point() +
           scale_color_manual(values = c(color.list,color.list)) +
           facet_wrap("Label") +
           theme_classic()
```

```{r}
seuratSC@meta.data <- cbind(seuratSC@meta.data,df[,c("dbImm_labels","moImm_labels")])
```

```{r}
DimPlot(seuratSC,
        reduction = "RNAumap",
        group.by = "dbImm_labels",
        cols = color.list)
```

# Multimodal Clustering based on WNN

```{r}
seuratSC <- seuratSC %>%
  FindMultiModalNeighbors(
    reduction.list = list("pca", "ADTpca"), 
    dims.list = list(1:20, 1:20),
    modality.weight.name = "RNA.weight"
  ) %>%
  FindClusters(
    graph.name = "wsnn",
    algorithm = 3,
    resolution = 0.5,
    verbose = FALSE) %>%
  RunUMAP(
    nn.name = "weighted.nn",
    reduction.name = "wnn.umap",
    reduction.key = "wnnUMAP_")

```

```{r}
DimPlot(seuratSC, reduction = 'wnn.umap', label = TRUE, repel = TRUE, label.size = 2.5,cols = color.list)
```

# Find Markers for each cluster

```{r}

Idents(seuratSC) <- "RNA_snn_res.0.5"
myMarkers <- seuratSC %>%
  FindAllMarkers(
    assay = "RNA",
    slot = "data"
  )
```

```{r fig.height=14}
myTopMarkers <- myMarkers %>%
  group_by(cluster) %>%
  filter(p_val_adj < 0.05) %>%
  top_n(25,wt = avg_log2FC)

seuratSC %>%
  ScaleData(assay = "RNA",
            features = myTopMarkers$gene) %>%
  DoHeatmap(
    features = myTopMarkers$gene,
    group.by = "RNA_snn_res.0.5",
    group.colors = color.list,
    # slot = "data",
    assay = "RNA",
    draw.lines = T
  )
```
